bayesflow/distributions/diagonal_student_t.py

```diff
@@ -1,40 +1,121 @@
+
         self.loc = loc
-        self.scale = scale
+        self.scale_diag = scale
         self.df = df
-        self.dim = None
         self.log_norm_const = None
+        self.dynamic_params = dynamic_params
+        self.seed_generator = seed_generator or keras.random.SeedGenerator()
 
     def sample(self, batch_shape: Shape) -> Tensor:
-        dist = scipy_student_t(df=self.df, loc=self.loc, scale=self.scale)
-        return dist.rvs(size=batch_shape)
+        dist = scipy_student_t(df=self.df, loc=self.loc, scale=self.scale_diag)
+        return dist.rvs(size=batch_shape + (self.dim,))
 
     def log_prob(self, tensor: Tensor) -> Tensor:
-        # TODO
-        pass
+        log_unnorm_pdf = self._log_unnormalized_prob(tensor)
+        return self.log_norm_const + log_unnorm_pdf
 
     def build(self, input_shape: Shape) -> None:
-        # TODO
-        pass
+        dim = int(input_shape[-1])
+
+        # At this point, loc and scale are either vectors or scalars, so it suffices to check any of them
+        if isinstance(self.loc, np.ndarray):
+            if self.loc.shape[0] != dim:
+                raise ValueError(
+                    f"The dimensions of your inference variables should be {self.loc.shape[0]} "
+                    f"but are currently {dim}. Please, ensure that dimensions match."
+                )
+        else:
+            self.loc = np.full(dim, self.loc)
+            self.scale_diag = np.full(dim, self.scale_diag)
+
+        self.dim = dim
+
+        # Pre-compute log normalizing constant
+        log_norm_const = -0.5 * self.dim * math.log(self.df * math.pi)
+        log_norm_const += math.lgamma(0.5 * (self.df + self.dim))
+        log_norm_const -= math.lgamma(0.5 * self.df)
+        log_norm_const -= np.sum(np.log(self.scale_diag))
+
+        # Add distribution parameters as keras weights
+        self.log_norm_const = self.add_weight(
+            name="log_norm_const",
+            shape=(1,),
+            trainable=False,
+            dtype="float32",
+            initializer=keras.initializers.Constant(log_norm_const),
+        )
+        self.df = self.add_weight(
+            name="dof", shape=(1,), trainable=False, dtype="float32", initializer=keras.initializers.Constant(self.df)
+        )
+        self.loc = self.add_weight(
+            name="loc",
+            shape=(self.dim,),
+            trainable=self.dynamic_params,
+            dtype="float32",
+            initializer=keras.initializers.Constant(self.loc),
+        )
+        self.scale_diag = self.add_weight(
+            name="scale_diag",
+            shape=(self.dim,),
+            trainable=self.dynamic_params,
+            dtype="float32",
+            initializer=keras.initializers.Constant(self.scale_diag),
+        )
 
     def _log_unnormalized_prob(self, tensor: Tensor) -> Tensor:
-        # TODO
-        pass
+        log_unnormalized_pdf = -0.5 * (self.df + self.dim)
+        log_unnormalized_pdf *= ops.log(
```

Comment:
This computation fails in torch backend due to invalid broadcasting. Splitting it into multiple lines would help narrow the error (and look a bit cleaner).